### PR TITLE
Added design doc in README

### PR DESCRIPTION
**Pull request by alexpitsikoulis**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,166 @@
...
+    Stop(processID string) (err error)
+    // Query a process to check its status
+    //    - processID: Process identifier (Process.ID)
+    // Returns status of specified process and error if an error was encountered
+    Query(processID string) (status Status, err error)
+    // Streams the process output
+    //    - ctx: context to cancel the log stream
+    //    - processID: Process identifier (Process.ID)
+    // Returns a chan to stream output to stdout/stderr and an error if an error was encountered
+    Stream(ctx context.Context, processID string) (logchan chan string, err error)
```
Comment (Joerger) [inline README.md:63]: What happens if the log stream encounters an error? Would the stream `chan string` just close with no explanation?

Comment (Joerger) [inline README.md:63]: nit: returning a string channel seems a bit odd. As a user of the library, it'd feel a bit awkward to interface with it.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,166 @@
...
+  rpc Start(StartRequest) returns (StartResponse);
+  rpc Stop(StopRequest) returns (StopResponse);
+  rpc Query(QueryRequest) returns (QueryResponse);
+  rpc Stream(StreamRequest) returns (stream StreamResponse);
+}
+```
+
+## CLI	
+
+A command line interface which allows the user to schedule Linux processes via ```nine2five```. The CLI allows communication between the user and the library using gRPC to ping the API.
```
Comment (Joerger) [inline README.md:119]: Can you add UX details for the CLI?

Comment (Joerger) [inline README.md:119]: > The CLI allows communication between the user and the library using gRPC to ping the API.

I'm a bit confused by the wording here, can you elaborate?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,166 @@
...
+TLS provides masking and data integrity during communication between the CLI and the API.
+
+### Authentication
+
+mTLS authentication will be used. Using openssl, the following assets will be generated for use in the auth schema.
+
+* Server
+    * CA private key and self-signed certificate
+    * Private key and certificate signing request (CSR)
+    * Signed certificate, based on self CA private key and self CSR
```
Comment (Joerger) [inline README.md:134]: What do you mean by "self" CA private key and "self" CSR? here and below.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,166 @@
...
+* Client
+    * CA private key and self-signed certificate
```
Comment (Joerger) [inline README.md:136]: Does the client have a separate CA from the server? If so, why?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,166 @@
...
+    * Private key and certificate signing request (CSR)
+    * Signed certificate, based on self CA private key and self CSR
+
+The certificate signature is checked, and the authentication attempts to find a CA certificate whose subject field matches the issuer field of the target certificate. If such a certificate is found the validator checks the signature on the target certificate using the public key in the CA certificate. No connection will be established if this check fails. CLI client and API both perform this check in order to validate one another without the use of intermediate certificates.
```
Comment (Joerger) [inline README.md:140]: This is a bit hard to parse, I think because you are mixing high level and low level explanations into a small excerpt. Can you try rewriting this, just encompassing the most crucial elements of an mTLS handshake? AKA what the client and server need to have and do to authenticate each other.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,166 @@
...
+    * stream
+* write
+    * query
+    * stream
+    * start
+    * stop
+
+The X.509 v3 extensions will be used to add the user role to the certificate. For that, the extension attribute roleOid 1.2.840.10070.8.1 = ASN1:UTF8String, must be requested in the Certificate Signing Request (CSR), when the user certificate is created, the user roles must be informed when the CA signs the the CSR. The information after UTF8String: is encoded inside of the x509 certificate under the given OID.
+
+#### gRPC interceptors
```
Comment (Joerger) [inline README.md:157]: What is this section meant to convey?

File: README.md [hunk 7]
```diff
@@ -1,2 +1,166 @@
...
+
+## CLI	
+
+A command line interface which allows the user to schedule Linux processes via ```nine2five```. The CLI allows communication between the user and the library using gRPC to ping the API.
+
+## Security
+
+### Transport
+
+TLS provides masking and data integrity during communication between the CLI and the API.
```
Comment (Joerger) [inline README.md:125]: Can you add information on what TLS version and Cipher suites will be supported?

File: README.md (hunk 6 above)
Comment (alexpitsikoulis) [reply README.md:251]: This may be scope creep, moving this to future improvements. This is in reference to limiting certain operations to certain user types and access restrictions.

Comment (alexpitsikoulis) [reply README.md]: Ok, I put it back

File: README.md [hunk 8]
```diff
@@ -1,2 +1,301 @@
...
+The library uses cgroups to limit CPU, memory, and disk IO resources. The cgroup limitations will be applied by creating cgroups in the Linux filesystem. For example starting a process will create a cgroup file at ```/sys/fs/cgroup/{subsytem}/{command_name}_{process_id}/{resource}``` (for the sake of example we will say ```subsystem``` is ```memory``` and ```resource``` is ```memory.limit_in_bytes```) and echo the speicified value into the file. The process command is then instantiated in a stopped state by first creating the command using the ```exec.Command``` function then setting ```Pdeathsig``` to ```syscall.SIGSTOP``` in the ```SysProcAttr``` field of the ```Cmd``` struct and then starting the command using ```{cmd}.Start```. The process is then added to the cgroup by appending the PID of the returned ```exec.Cmd``` struct to the ```/sys/fs/cgroup/memory/{command_name}_{process_id}/tasks``` file. Afterwards the process is started with ```syscall.Kill({pid}, syscall.SIGCONT)```. This will start the process with the cgroup limits. After the process exits the cgroup directory will be deleted to avoid buildup of unused cgroups.
```
[the remark this answers is not in the document]
Comment (alexpitsikoulis) [reply README.md:18]: It sends the SIGSTOP on the death of the creation thread for the process

File: README.md [hunk 9]
```diff
@@ -1,2 +1,301 @@
...
+type Command struct {
+    // Base command name
+    Name string
+    // List of arguments to the command
+    Args []string
+}
+
+// Process represents the process and its associated status data
+type Process struct {
+    // Unique ID
```
Comment (codingllama) [inline README.md:45]: This is not the PID, correct?

What is your ID of choice, then?

File: README.md [hunk 10]
```diff
@@ -1,2 +1,301 @@
...
+    // Status of the process.
+    Status *Status
+}
+
+type State int
+const (
+    Running ProcessState = iota
+    Stopped
+    Exited
+    Killed
```
Comment (codingllama) [inline README.md:58]: What is the difference between these? Do we need all of them?

File: README.md [hunk 11]
```diff
@@ -1,2 +1,301 @@
...
+// Command is the body of a request to start a process as received from the API or CLI
+type Command struct {
+    // Base command name
+    Name string
+    // List of arguments to the command
+    Args []string
+}
+
+// Process represents the process and its associated status data
+type Process struct {
```
Comment (codingllama) [inline README.md:44]: Should this be private? I noticed we never get a Process from the Worker (which is fine by me).

File: README.md [hunk 12]
```diff
@@ -1,2 +1,301 @@
...
+  repeated string args = 2;
+}
+
+message StartResponse {
+  string processID = 1;
+}
+
+message StopRequest {
+  string processID = 1;
+  bool gracefully = 2
```
Comment (codingllama) [inline README.md:113]: nit: This breaks syntax highlighting for the message below. (Super minor ofc.)

```suggestion
  bool gracefully = 2;
```

Comment (codingllama) [inline README.md:113]: I don't think the graceful option is present in the API. Drop this field?

File: README.md [hunk 13]
```diff
@@ -1,2 +1,301 @@
...
+syntax = "proto3";
+option go_package = "github.com/alexpitsikoulis/go-job-worker/nine2five/internal/proto";
+
+message StartRequest {
+  string name = 1;
+  repeated string args = 2;
+}
+
+message StartResponse {
+  string processID = 1;
```
Comment (codingllama) [inline README.md:108]: nit: proto style uses lower_camel, like process_id. (Not a big deal.)

https://protobuf.dev/programming-guides/style/#message-field-names

File: README.md [hunk 14]
```diff
@@ -1,2 +1,301 @@
...
+}
+
+message QueryResponse {
+  int32 pid = 1;
+  int32 exitCode = 2;
+  int32 state = 3;
+}
+
+message StreamRequest {
+  repeated jobID = 1;
```
Comment (codingllama) [inline README.md:130]: Should this be processID too? (Or process_id.)

File: README.md [hunk 15]
```diff
@@ -1,2 +1,301 @@
...
+        * -p, --port (optional; if not provided will default to 8080)
+    * Output
+        * Success message: In case of successful server start, a success message will be displayed
+            Example: ```Server running on port {port}```
+        * Error: In case of failed server start, the error will be displayed
+            Example: ```Failed to start server: {error_message}```
+    Example commands:
+        * ```go-job-worker-server start```
+        * ```go-job-worker-server start -p 8009```
+* stop
```
Comment (codingllama) [inline README.md:156]: Why does the server need stop and kill subcommands?

File: README.md [hunk 16]
```diff
@@ -1,2 +1,301 @@
...
+3. Generate Client Certificate
+    1. Generate key and save to ```client.key``` using ```openssl genrsa -out client.key 4096```
+    2. Create CSR and save to ```client.csr``` using ```openssl req -new -key client.key -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN=github.com/alexpitsikoulis" -out client.csr```
+    3. Sign client CSR and save signed certificate to ```client.crt``` using ```openssl x509 -req -in client.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client.crt -days 825 -sha256```
+
+Both the client and server must possess and present their own certificates. During the authentication process, each party verifies the other’s certificate. This is done by checking if the certificate was issued by a trusted Certificate Authority (CA). They do this by comparing the issuer of the presented certificate against their list of trusted CA certificates. If a match is found, the signature on the presented certificate is validated using the public key of the corresponding CA certificate. A successful verification establishes a trusted connection. If either the client or server cannot validate the other's certificate, the connection will not be established.
```
Comment (codingllama) [inline README.md:274]: How are the certificates supplied to the server and client?

File: README.md [hunk 17]
```diff
@@ -1,2 +1,301 @@
...
+* read
+    * query
+    * stream
+* write
+    * query
+    * stream
+    * start
+    * stop
+
+The X.509 v3 extensions will be used to add the user role to the certificate. For that, the extension attribute roleOid 1.2.840.10070.8.1 = ASN1:UTF8String, must be requested in the Certificate Signing Request (CSR), when the user certificate is created, the user roles must be informed when the CA signs the the CSR. The information after UTF8String: is encoded inside of the x509 certificate under the given OID.
```
Comment (codingllama) [inline README.md:289]: What is the string format for the extension?

File: README.md [hunk 18]
```diff
@@ -1,2 +1,301 @@
...
+    3. Sign client CSR and save signed certificate to ```client.crt``` using ```openssl x509 -req -in client.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client.crt -days 825 -sha256```
+
+Both the client and server must possess and present their own certificates. During the authentication process, each party verifies the other’s certificate. This is done by checking if the certificate was issued by a trusted Certificate Authority (CA). They do this by comparing the issuer of the presented certificate against their list of trusted CA certificates. If a match is found, the signature on the presented certificate is validated using the public key of the corresponding CA certificate. A successful verification establishes a trusted connection. If either the client or server cannot validate the other's certificate, the connection will not be established.
+
+### Authorization
+The CLI client's certificate will be extended with the user roles which will allow the API to grant or deny access based on the user's role and the action they are trying to perform.
+
+Roles -> Privileges:
+* read
```
Comment (codingllama) [inline README.md:280]: Read and write roles are a simple system, but it's not a useful one. For example, a user with write has pretty much unlimited power over the system, being able to read and interfere with anyone's jobs.

Do you have any ideas for an authn system that is still simple, but generally more useful from an UX perspective?

File: README.md [hunk 19]
```diff
@@ -1,2 +1,301 @@
...
+
+The process for generating the certificate and keys is as follows:
+1. Generate CA
+    1. A 4096-bit RSA key is generated and saved to ```ca.key``` using ```openssl genrsa -out ca.key 4096```
+    2. A self-signed root CA certificate is generated with the key at ```ca.key``` and saved to ```ca.crt``` using ```openssl req -new -x509 -key ca.key -sha256 -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN=github.com/alexpitsikoulis" -days 1024 -out ca.crt```
+2. Generate Server Certificate
+    1. Generate key and save to ```server.key``` using ```openssl genrsa -out server.key 4096```
+    2. Create CSR and save to ```server.csr``` using ```openssl req -new -key server.key -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN=github.com/alexpitsikoulis" -out server.csr```
+    3. Sign server CSR and save signed certificate to ```server.crt``` using ```openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 825 -sha256```
+3. Generate Client Certificate
```
Comment (codingllama) [inline README.md:269]: How can we distinguish different clients/users?

File: README.md [hunk 20]
```diff
@@ -1,2 +1,301 @@
...
+
+#### Certificates
+* X.509
+* Signature Algorithm: sha256WithRSAEncryption
+* Public Key Algorithm: rsaEncryption
+* RSA Public-Key: (4096 bit)
+* roleOid 1.2.840.10070.8.1 = ASN1:UTF8String (for the client certificate)
+
+## Future Improvements
+* Implement command filtering - When a request is received it will be parsed to ensure no dangerous commands are being executed, if a process is dangerous or malicious an error will be returned.
```
Comment (codingllama) [inline README.md:299]: How would you implement this?

File: README.md [hunk 21]
```diff
@@ -1,2 +1,263 @@
...
+Prototype job worker service that provides an API to run arbitrary Linux processes.
+
+## Overview
+
+### Library (nine2five)
+
+A generic, reusable Go package that acts as a worker to exexcute, monitor, and stop arbitrary Linux processes. This library is responsible for starting and stopping processes as well as streaming process output and handling process errors.
+This library will use a map in local memory to store the statuses of running, completed, and failed jobs.
+Processes are started by initializing a new Linux process using the ```exec.Cmd``` struct. Processes are handled in parallel by dedicated ```Worker``` nodes per process.
+The library will stream the output of the job to a log file. The log files' default location will be ```/tmp/{command_name}_{process_id}.log```. It will do so by reading from the process' stdout and piping it into the log file as bytes. As this will cause a buildup of old log files, eventually a new stragey will be implemented to combat that. Whether that will be purging based on expiry by age or by disk space usage is yet to be determined. Another approach to be considered is the use of S3 buckets.
```
[the remark this answers is not in the document]
Comment (codingllama) [reply README.md:15]: +1

File: README.md [hunk 22]
```diff
@@ -1,2 +1,301 @@
...
+}
+
+message QueryRequest {
+  string processID = 1;
+}
+
+message QueryResponse {
+  int32 pid = 1;
+  int32 exitCode = 2;
+  int32 state = 3;
```
Comment (codingllama) [inline README.md:126]: What are the possible states? Would an enum be a good fit here?

File: README.md [hunk 23]
```diff
@@ -1,2 +1,166 @@
...
+  int32 exitCode = 2;
+  bool exited = 3;
+}
+
+message StreamRequest {
+  string jobID = 1;
+}
+
+message StreamResponse {
+  string output = 1;
```
[the remark this answers is not in the document]
Comment (codingllama) [reply README.md:134]: +1

File: README.md [hunk 24]
```diff
@@ -1,2 +1,301 @@
...
+The library will stream the output of the job to a log file. The log files' default location will be ```/tmp/{command_name}_{process_id}.log```. It will do so by reading from the process' stdout and piping it into the log file as bytes. As this will cause a buildup of old log files, eventually a new stragey will be implemented to combat that. Whether that will be purging based on expiry by age or by disk space usage is yet to be determined. Another approach to be considered is the use of S3 buckets.
+Upon requests to the API to stream output from a process, the content from the log file will be read to the terminal, tailing the logs as a process is running, seeing the logs in real time. This is achieved by reading the bytes of the file to stdout/stderr, when the end of the file is reached it will then wait for filesystem notifications (listened for by the ```fsnotify``` package) regarding changes to that log file. Streaming then resumes to output the new bytes and waits again. In the case of an unreadable byte, a placeholder character ```�``` will be used in its place. This continues until the process exits, at which point the stream loop is closed and the reader destroyed.
```
Comment (codingllama) [inline README.md:16]: > listened for by the ```fsnotify``` package

Is this the package you mean? https://pkg.go.dev/github.com/fsnotify/fsnotify.

File: README.md [hunk 25]
```diff
@@ -1,2 +1,301 @@
...
+
+        * Error: In case of failed query, the error will be displayed
+            Example: ```Failed to query status of process df685efd-a69a-4527-b2ec-5deaf0309027```
+    
+    Example command: ```go-job-worker query df685efd-a69a-4527-b2ec-5deaf0309027```
+* stream
+    * Args
+        * Process ID: ID of process whose output is to be streamed (required)
+    * Output
+        * Process output: If the ```-tail``` flag was provided the live output of the process will be streamed
```
Comment (codingllama) [inline README.md:216]: How is the --tail flag implemented? I don't see any toggles in the RPC.

File: README.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (alexpitsikoulis) [reply README.md:134]: I addressed this by using the placeholder character for unreadable characters and monitoring resource usage to kill long running processes that are using a lot of resources or spawning a lot of subprocesses

File: README.md (hunk 25 above)
Comment (alexpitsikoulis) [reply README.md:216]: Ahh missed this in cleanup when i removed the --tail flag. Will remove that bit.

File: README.md (hunk 24 above)
Comment (alexpitsikoulis) [reply README.md:16]: Yes it is

File: README.md (hunk 17 above)
Comment (alexpitsikoulis) [reply README.md:289]: 1.2.840.10070.8.1 = ASN1:UTF8String:[user_role]

Comment (codingllama) [reply README.md:342]: So, single role string always, correct?

File: README.md [hunk 26]
```diff
@@ -1,2 +1,354 @@
...
+  ProcessState state = 3;
+}
+
+message StreamRequest {
+  string process_id = 1;
+}
+
+message StreamResponse {
+  // The log file will be streamed as bytes and interpreted in a display-safe format by the CLI
+  repeated int32 output = 1;
```
Comment (codingllama) [inline README.md:154]:
```suggestion
  bytes output = 1;
```

https://protobuf.dev/programming-guides/proto3/#scalar

File: README.md [hunk 27]
```diff
@@ -1,2 +1,354 @@
...
+2. Generate Server Certificate
+    1. Generate key and save to ```server.key``` using ```openssl genrsa -out server.key 4096```
+    2. Create CSR and save to ```server.csr``` using ```openssl req -new -key server.key -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN=github.com/alexpitsikoulis" -out server.csr```
+    3. Sign server CSR and save signed certificate to ```server.crt``` using ```openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 825 -sha256```
+3. Generate Client Certificate
+    1. Generate key and save to ```client.key``` using ```openssl genrsa -out client.key 4096```
+    2. Create CSR and save to ```client.csr``` using ```openssl req -new -key client.key -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN={unique ID per client registration}" -out client.csr```
+    3. Sign client CSR and save signed certificate to ```client.crt``` using ```openssl x509 -req -in client.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client.crt -days 825 -sha256```
+
+The process for supplying the certificate to the server will be as follows:
```
Comment (codingllama) [inline README.md:280]: No need for the Go code, just saying that it looks for certain files in its workdir is good enough.

File: README.md [hunk 28]
```diff
@@ -1,2 +1,354 @@
...
+caCertPool := x509.NewCertPool()
+caCertPool.AppendCertsFromPEM(caCert)
+&tls.Config{
+    Certificates: []tls.Certificate{cert},
+    ClientCAs: caCertPool,
+    ClientAuth: tls.RequireAndVerifyClientCert,
+}
+```
+
+The process for supplying the certificate to the client will be as follows:
```
Comment (codingllama) [inline README.md:301]: Same here.

File: README.md [hunk 29]
```diff
@@ -1,2 +1,354 @@
...
+    2. A self-signed root CA certificate is generated with the key at ```ca.key``` and saved to ```ca.crt``` using ```openssl req -new -x509 -key ca.key -sha256 -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN=github.com/alexpitsikoulis" -days 1024 -out ca.crt```
+2. Generate Server Certificate
+    1. Generate key and save to ```server.key``` using ```openssl genrsa -out server.key 4096```
+    2. Create CSR and save to ```server.csr``` using ```openssl req -new -key server.key -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN=github.com/alexpitsikoulis" -out server.csr```
+    3. Sign server CSR and save signed certificate to ```server.crt``` using ```openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 825 -sha256```
+3. Generate Client Certificate
+    1. Generate key and save to ```client.key``` using ```openssl genrsa -out client.key 4096```
+    2. Create CSR and save to ```client.csr``` using ```openssl req -new -key client.key -subj "/C=US/ST=California/L=LosAngeles/O=go-job-worker/CN={unique ID per client registration}" -out client.csr```
```
Comment (codingllama) [inline README.md:277]: > unique ID per client registration

How are the unique IDs determined?

What is a "client registration"?

File: README.md [hunk 30]
```diff
@@ -1,2 +1,354 @@
...
+```
+
+
+Both the client and server must possess and present their own certificates. During the authentication process, each party verifies the other’s certificate. This is done by checking if the certificate was issued by a trusted Certificate Authority (CA). They do this by comparing the issuer of the presented certificate against their list of trusted CA certificates. If a match is found, the signature on the presented certificate is validated using the public key of the corresponding CA certificate. A successful verification establishes a trusted connection. If either the client or server cannot validate the other's certificate, the connection will not be established.
+
+### Authorization
+The CLI client's certificate will be extended with the user roles which will allow the API to grant or deny access based on the user's role and the action they are trying to perform.
+
+Roles -> Privileges:
+* admin
```
Comment (codingllama) [inline README.md:328]: I like the new roles better, makes much sense.

My 2c is that I would limit the set to just "admin" and "write". "readonly-admin" is nice to have, but not paramount. "read" can't start jobs, so I don't see its powers having a purpose (unless we get into role change scenarios, but let's not ;) ).

File: README.md [hunk 31]
```diff
@@ -1,2 +1,354 @@
...
+    * all privileges over all jobs
+* readonly-admin
+    * query (all jobs)
+    * stream (all jobs)
+* read
+    * query (own jobs)
+    * stream (own jobs)
+* write
+    * query (own jobs)
```
Comment (codingllama) [inline README.md:337]: How is job ownership mapped into the system? It's not mentioned apart from here.

File: README.md (hunk 20 above)
Comment (codingllama) [reply README.md]: Re-opened, I don't see a reply.

File: README.md (hunk 29 above)
Comment (alexpitsikoulis) [reply README.md:277]: It will just be a generated UUID, client registration meaning per cert creation.

Comment (alexpitsikoulis) [reply README.md:277]: Ahh nevermind I see the issue, when a cert expires that user loses access to their jobs.

Comment (codingllama) [reply README.md:277]: You could mint a new certificate with the same UUID, so I don't see the issue. You could also use descriptive names, emails, etc - that's all fine. I just wanted to clarify what the plan is.

File: README.md (hunk 31 above)
Comment (alexpitsikoulis) [reply README.md:290]: I have made ```Process``` public and added an OwnerID field. ```Start``` now returns a ```Process``` and this is how I will handle ownership of jobs.

File: README.md (hunk 30 above)
Comment (alexpitsikoulis) [reply README.md:287]: Removed

File: README.md (hunk 29 above)
Comment (alexpitsikoulis) [reply README.md]: You're right, was overthinking this one. I will use UUID and mint new certificates with the same UUID.

File: README.md (hunk 28 above)
Comment (alexpitsikoulis) [reply README.md]: Removed

File: README.md (hunk 27 above)
Comment (alexpitsikoulis) [reply README.md]: Removed

File: README.md (hunk 26 above)
Comment (alexpitsikoulis) [reply README.md]: Changed

File: README.md [hunk 32]
```diff
@@ -1,2 +1,354 @@
...
+message QueryRequest {
+  string process_id = 1;
+}
+
+enum ProcessState {
+    UnknownState = 0,
+    Running = 1,
+    Stopped = 2,
+    Exited = 3,
+    Killed = 4
```
[the remark this answers is not in the document]
Comment (alexpitsikoulis) [reply README.md]: Missed this when changing the Go enum. Removed extra entries

File: README.md [hunk 33]
```diff
@@ -1,2 +1,354 @@
...
+The library will stream the output of the job to a log file. The log files' default location will be ```/tmp/{command_name}_{process_id}.log```. It will do so by reading from the process' stdout and stderr and piping it into the log file as bytes. Both stdout and stderr will be assigned to the same filewriter and write to the same file in chronological order. As this will cause a buildup of old log files, eventually a new stragey will be implemented to combat that. Whether that will be purging based on expiry by age or by disk space usage is yet to be determined. Another approach to be considered is the use of S3 buckets.
```
[the remark this answers is not in the document]
Comment (alexpitsikoulis) [reply README.md]: I took a new approach using 2 buffered channels. One for stdout and one for stderr that receive output one line at a time and are piped into the log file from the same goroutine.

File: README.md (hunk 20 above)
Comment (alexpitsikoulis) [reply README.md]: Added details on possible implementation paths (regex, command blacklisting, records of commands that routinely exhibit problematic behavior or anomalies in fs/system/network monitoring.

File: README.md [hunk 34]
```diff
@@ -1,2 +1,301 @@
...
+Prototype job worker service that provides an API to run arbitrary Linux processes.
+
+## Overview
+
+### Library (nine2five)
+
+A generic, reusable Go package that acts as a worker to exexcute, monitor, and stop arbitrary Linux processes. This library is responsible for starting and stopping processes as well as streaming process output and handling process errors.
+This library will use a map in local memory to store the statuses of running, completed, and failed jobs.
+Processes are started by initializing a new Linux process using the ```exec.Cmd``` struct. Processes are handled in parallel by dedicated ```Worker``` nodes per process.
+These processes will be monitored for excessive resource (CPU, disk IO, memory) usage (this will be monitored using cgroups), unusual patters (subprocess spawning, system calls, attempts to access restricted resources), filesystem changes, network activity, and prolonged execution time (particularly with large resource usage) to prevent unsafe, hanging, or unpredictable commands.
```
[the remark this answers is not in the document]
Comment (alexpitsikoulis) [reply README.md]: Implementation details added and this functionality was moved to future improvements.

File: README.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (alexpitsikoulis) [reply README.md]: Addressed by streaming bytes rather than text.
